### PR TITLE
Support for remaining subquery queries in Gen4

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description
This PR adds support for the remaining queries with subqueries in Gen4

## Related Issue(s)
#7280 

## Checklist
- [ ] Should this PR be backported?
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 1]
```diff
@@ -152,6 +147,80 @@ func optimizeSubQuery(ctx planningContext, op *abstract.SubQuery) (queryTree, er
...
+						}
+						columnIndex := columnIndexes[0]
+						outerTree.vars[bindVar] = columnIndex
+						return false
+					}
+				}
+				return true
+			}, nil)
+
+			return rt, rewriteError
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/route_planning.go:199]: We should move this to a non-exported method like we did for `mergeSubQuery`

**Review by frouioui — APPROVED** [review]

File: go/vt/vtgate/planbuilder/rewrite.go [hunk 2]
```diff
@@ -112,18 +113,58 @@ func queryRewrite(ctx planningContext, statement sqlparser.SelectStatement) erro
...
 	switch node := cursor.Node().(type) {
 	case *sqlparser.Select:
 		rewriteHavingClause(node)
 	case *sqlparser.ExistsExpr:
 		return r.rewriteExistsSubquery(cursor, node)
 	case *sqlparser.Subquery:
+		r.isInSubquery++
 		rewriteSubquery(cursor, r, node)
+	case *sqlparser.AliasedTableExpr:
+		if r.isInSubquery == 0 {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/rewrite.go:130]: nit: squash this and the next if into a single one with `||` between

File: go/vt/vtgate/planbuilder/route_planning.go [hunk 3]
```diff
@@ -152,6 +147,80 @@ func optimizeSubQuery(ctx planningContext, op *abstract.SubQuery) (queryTree, er
...
+			if err != nil {
+				return nil, err
+			}
+			otherTree := outerTree.lhs
+			if a.tableID() == outerTree.lhs.tableID() {
+				otherTree = outerTree.rhs
+			}
+
+			var rewriteError error
+			sqlparser.Rewrite(subQueryInner.SelectStatement.Where.Expr, func(cursor *sqlparser.Cursor) bool {
```
Comment (systay) [inline go/vt/vtgate/planbuilder/route_planning.go:175]: a few sprinkled comments should make it easier to understand

**Review by systay — APPROVED** [review]

File: go/vt/vtgate/planbuilder/rewrite.go [hunk 4]
```diff
@@ -112,18 +113,58 @@ func queryRewrite(ctx planningContext, statement sqlparser.SelectStatement) erro
...
+		}
+		tableName, isNotDerived := node.Expr.(sqlparser.TableName)
+		if !isNotDerived {
+			break
+		}
+
+		tableSet := r.semTable.TableSetFor(node)
+		if tableSet == 0 {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/rewrite.go]: no no, this was just fail-safe code, but it is okay to remove

File: go/vt/vtgate/planbuilder/rewrite.go [hunk 5]
```diff
@@ -112,18 +113,58 @@ func queryRewrite(ctx planningContext, statement sqlparser.SelectStatement) erro
...
 	case *sqlparser.Subquery:
+		r.isInSubquery++
 		rewriteSubquery(cursor, r, node)
+	case *sqlparser.AliasedTableExpr:
+		if r.isInSubquery == 0 {
+			break
+		}
+		if !node.As.IsEmpty() {
+			break
+		}
```
[the remark this answers is not in the document]
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/rewrite.go]: Yes good catch

File: go/vt/vtgate/planbuilder/rewrite.go (hunk 2 above)
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/rewrite.go:130]: done

File: go/vt/vtgate/planbuilder/route_planning.go (hunk 1 above)
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/route_planning.go:199]: done

File: go/vt/vtgate/planbuilder/route_planning.go (hunk 3 above)
Comment (GuptaManan100) [reply go/vt/vtgate/planbuilder/route_planning.go]: added

**Review by systay — APPROVED** [review]

**Pull request merged**